text/0000-gem-exec.md

```diff
@@ -0,0 +1,95 @@
+    -v, --version VERSION            Specify version of gem to exec
+        --[no-]prerelease            Allow prerelease versions of a gem
+                                     to be installed
+    -g, --gem GEM                    run the executable from the given gem
+
+
+  Install/Update Options:
+        --conservative               Prefer the most recent installed version,
+                                     rather than the latest version overall
+
+
+  Common Options:
+    -h, --help                       Get help on this command
+    -V, --[no-]verbose               Set the verbose level of output
+    -q, --quiet                      Silence command progress meter
+        --silent                     Silence RubyGems output
+        --config-file FILE           Use this config file instead of default
+        --backtrace                  Show stack backtrace on errors
+        --debug                      Turn on Ruby debugging
+        --norc                       Avoid loading any .gemrc file
+
+
+  Arguments:
+    COMMAND  the executable command to run
+
+  Summary:
+    Run a command from a gem
+
+  Description:
+
+  Defaults:
+    --version '>= 0'
+```
+
+# Reference-level explanation
+
+A new `gem` subcommand called `exec` will be added to RubyGems.
+
+`gem exec` will use its own gem path to install gems (when needed), and will attempt to load the requested binary from the existing set of gems.
+If the binary is not found, the command will install the gem (by default of the same name as the binary, though it can be set by a command line option),
+and then activate and execute that binary.
+
+# Drawbacks
+
+This proposal should have no impact on existing users, since it is purely additive.
+
+One potential area of confusion is with Bundler / users using Gemfiles.
+To start, we plan on ignoring the Gemfile entirely when running `gem exec`, which may be confusing if the user is attempting to use the Gemfile to customize gem sources.
+Additionally, `gem exec` is confusable with the `bundle exec` command, and they will work differently. A potential way to ameliorate that confusion is for `gem exec` to check if the requested executable exists in the current bundle, and advise the user to use `bundle exec` in that case.
+
+# Rationale and Alternatives
+
+We could continue to point users to the existing [`gemx`](https://github.com/rubygems/gemx) gem to solve this problem, or encourage using `gem install $gem && $gemBinary`, but those solutions (which have existed for years) are not sufficiently discoverable.
+
+Implementing this proposal will make it easy for users to run one-off commands that _happen_ to be shipped as gems, and instructions to `gem exec $gem` can become widespread.
+
+# Unresolved questions
+
+Should there be an option to use an isolated gem path for `gem exec` installs? Should it be required or the default? Should we look into using existing gems but not installing into the default path?
```

Comment:
> Should we look into using existing gems

What is actual plan for this? Let's say there is locally installed `rails 7.0.0` and `gem exec rails -r "~> 7.0` is requested. There is also newer version available remotely. Is it going to first install latest version matching related constraint and run `rails` command in there or is it going to be satisfied with local already installed matching version?

Reply:
current plan is to not use existing gem unless `--conservative` is given

Reply:
I think it makes sense for the default usecase to be installing and then running the latest version of a package, and only skipping the install if you already have the newest version installed. If you want to run whatever you already have installed without installing the latest, you can probably run `rails new .` instead of `gem exec rails new .`, right?

Reply:
Given that we default to not using the already-installed gem, I think it's reasonable to have a separate path for installs without the conservative flag, so you can `gem exec` without filling up your global gemset.